compiler/rustc_mir/src/transform/large_enums.rs

```diff
@@ -0,0 +1,266 @@
+                        if max.bytes() - min.bytes() < D {
+                            return None;
+                        }
+                        let mut discr_sizes = vec![Size::ZERO; adt_def.discriminants(tcx).count()];
+                        for (var_idx, layout) in variants.iter_enumerated() {
+                            let disc_idx =
+                                adt_def.discriminant_for_variant(tcx, var_idx).val as usize;
+                            assert_eq!(discr_sizes[disc_idx], Size::ZERO);
+                            discr_sizes[disc_idx] = layout.size;
+                        }
+                        Some((variants.len() as u64, discr_sizes))
+                    }
+                }
+            }
+            _ => None,
+        }
+    }
+    fn optim(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        let mut alloc_cache = FxHashMap::default();
+        let body_did = body.source.def_id();
+        let (bbs, local_decls) = body.basic_blocks_and_local_decls_mut();
+        for bb in bbs {
+            bb.expand_statements(|st| {
+                match &st.kind {
+                    StatementKind::Assign(box (
+                        lhs,
+                        Rvalue::Use(Operand::Copy(rhs) | Operand::Move(rhs)),
+                    )) => {
+                        let ty = lhs.ty(local_decls, tcx).ty;
+
+                        let source_info = st.source_info;
+                        let span = source_info.span;
+
+                        let (num_variants, sizes) =
+                            if let Some(cand) = Self::candidate(tcx, ty, body_did) {
+                                cand
+                            } else {
+                                return None;
+                            };
+                        let adt_def = ty.ty_adt_def().unwrap();
+
+                        let alloc = if let Some(alloc) = alloc_cache.get(ty) {
+                            alloc
+                        } else {
+                            let data_layout = tcx.data_layout();
+                            let ptr_sized_int = data_layout.ptr_sized_integer();
+                            let mut data =
+                                vec![0; (ptr_sized_int.size().bytes() * num_variants) as usize];
+
+                            let mut curr = 0;
+                            for byte in sizes
+                                .iter()
+                                .flat_map(|sz| match data_layout.endian {
+                                    rustc_target::abi::Endian::Little => sz.bytes().to_le_bytes(),
+                                    rustc_target::abi::Endian::Big => sz.bytes().to_be_bytes(),
+                                })
+                                .take(data.len())
+                            {
+                                data[curr] = byte;
+                                curr += 1;
+                            }
```

Comment:
This always encodes the variant size using 8-bytes (`bytes()` returns `u64` type). It should encode them using as many bytes as `usize` type has on given target.

Reply:
I might need to rework this to make it less ugly but should be doing that now